Backend Refactor

sraboy opened the pull request:
This is regarding [angr issue 575](https://github.com/angr/angr/issues/575). 

I wanted you guys to see where I'm going with this from a refactoring perspective. So far, this does not break the external API as far as angr is concerned but does rename/move/remove some modules and classes. It also depends on the [corresponding archinfo PR](https://github.com/angr/archinfo/pull/14).

A summary of the changes thus far:
 - Eliminated `PE._handle_relocs` in favor of `__register_relocs` and `_make_reloc`. Most of the code was reused/refactored as required. Since those two functions are so similar to those in `ELF`, they might be able to get moved up to `Backend` if the hacks for MIPS (in `ELF`) and for DLL imports (in `PE`) can be done better.
- Added `relocations.generic_pe` module, which currently contains `WinReloc` (bit of a hack to handle DLLs).
- Added new modules and classes for each arch's relocations, like I did in PR #79. Combined with `generic_pe`, this eliminates the `pe.reloc` module. This required modules and arches to be named `pe<arch>` since `get_relocations` and `archinfo.defines` are OS-agnostic.
- Moved the various if-else checks for `IMAGE_REL_BASED_*` in `WinReloc` to their own relocation classes in `generic_pe`.

All these changes are for relocations. I also added an `align` member to `PESection` even though, in PEs, this is actually a property of the image. It's to maintain API compatibility with `ELFSection`.

I've got a bit more to do for relocations:
- [x] Clean up (remove some C-isms, logger names, remove commented-out junk)
- [ ] Make things more "Pythonic". My C background gets in the way here...
- [x] Move `WinReloc` to `backends.relocations` or, if possible, eliminate it entirely.
- [x] If it doesn't extensively blow things up, move `relocations/<arch>.py` to `relocations/elf` and `relocations/pe` subdirs and eliminate the `pe<arch>` naming.
- [x] Squash without losing context

And for other PE backend stuff:
- [ ] Add properties to the base `Section` class that ELF/PE can override (align, entsize, flags, type)
- [x] Move that `align` change to its own (to-be-squashed) commit for relevant changes
- [ ] Add parsing for PE's `characteristics` to get `type` and `flags`
- [x] Squash without losing context
- [ ] Deprecate `pe._pe` in favor of `pe.reader`

As a side note, I'll be losing my official "work time" for angr here shortly so my contributions will probably slow down but I'm on Slack and having fun so I hope to make time outside of work.

ltfish commented:
> remove some C-isms, logger names, remove commented-out junk

This is awesome.

ltfish commented on cle/backends/pe/pe.py:
```diff
@@ -62,7 +64,7 @@ def __init__(self, *args, **kwargs):
         self._handle_imports()
         self._handle_exports()
-        self._handle_relocs()
+        self.__register_relocs()
```
What's double underscore for? We rarely use such a naming style.

sraboy replied:
Just matching [ELF's](https://github.com/angr/cle/blob/08e9115033da89c0a4691b228def56f68948069d/cle/backends/elf/elf.py#L483). There's a few in there like that. If you'd rather eliminate them, I can update ELF's to go along with this PR.

(For some reason, this comment said it posted... but didn't.)

ltfish replied:
Oops, I don't remember those methods at all. Never mind.

rhelmot replied:
If I recall correctly, the elf method was double-underscored because it should NEVER be called outside of the context it is originally used, and attempting to do so will fuck everything up

ltfish commented on cle/backends/pe/pe.py:
```diff
@@ -121,9 +123,12 @@ def _handle_imports(self):
-                    reloc = WinReloc(self, symb, AT.from_lva(imp.address, self).to_rva(), entry.dll)
-                    self.imports[imp_name] = reloc
-                    self.relocs.append(reloc)
+                    addr=AT.from_lva(imp.address, self).to_rva()
```
Spaces around the `=` sign please.

ltfish commented on cle/backends/relocations/peamd64.py:
```diff
@@ -0,0 +1,21 @@
+
+l = logging.getLogger('cle.relocations.peamd64')
+
+arch = 'peAMD64'
```
Where is this `arch` string used?

sraboy replied:
Passed to [get_relocations](https://github.com/sraboy/cle/blob/3d17c5d6b648a4b14df66f1df505c9b3c37881a5/cle/backends/relocations/__init__.py#L40) by [elf backend](https://github.com/sraboy/cle/blob/3d17c5d6b648a4b14df66f1df505c9b3c37881a5/cle/backends/elf/elf.py#L548), so I was just matching it. It does feel a bit wonky since the module's name is the arch.

ltfish replied:
I might be misunderstanding it. Isn't `self.arch.name` passed to `get_relocations()`? How can you access the `arch` argument here in this file by `self.arch`?

If this is just descriptive, you probably want to change them to comment strings.

sraboy replied:
Sorry, I should have been more descriptive. It's not solely descriptive since it's [used as a dictionary key by relocations](https://github.com/sraboy/cle/blob/3d17c5d6b648a4b14df66f1df505c9b3c37881a5/cle/backends/relocations/__init__.py#L37).

I'm just passing in [`'pe' + self.arch.name`](https://github.com/sraboy/cle/blob/3d17c5d6b648a4b14df66f1df505c9b3c37881a5/cle/backends/pe/pe.py#L178) for now.

This can probably be eliminated if/when I move each of those arch-specific modules to `elf` and `pe` subdirectories; then the subdirectory can be used as a key to a dict wrapping the current `ALL_RELOCATIONS` dict. This will also allow for adding new executable formats.

Since relocations are generally specific to a backend, perhaps each backend module should have its own relocations subdir...

ltfish replied:
> Since relocations are generally specific to a backend, perhaps each backend module should have its own relocations subdir

ltfish replied:
> Since relocations are generally specific to a backend, perhaps each backend module should have its own relocations subdir

@rhelmot What do you think?

sraboy replied:
Similarly, `Relocation` has some ELF-specific things in it, like `is_rela` and `addend`. With separate subdirs within the backends, these things could be pieced out to an `ELFReloc`.

ltfish replied:
I agree. Let's see what @rhelmot says.

rhelmot replied:
Yes, I was going to suggest that we move the base relocations in to `cle/backends/relocation.py` and then the ELF ones into `cle/backends/elf/relocation/` and the PE ones into `cle/backends/pe/relocation{.py, /}`

rhelmot replied:
Then I guess get_relocation would be moved from the base into the specific module so we wouldn't have to namespace the arches like that. The "scraping" code could stay in the base, though.

ltfish commented on cle/backends/relocations/pearm.py:
```diff
@@ -0,0 +1,27 @@
+
+l = logging.getLogger('cle.relocations.pearm')
+
+arch = 'pearm'
```
Again, where is this `arch` string used?

sraboy replied:
Passed to [get_relocations](https://github.com/sraboy/cle/blob/3d17c5d6b648a4b14df66f1df505c9b3c37881a5/cle/backends/relocations/__init__.py#L40) by [elf backend](https://github.com/sraboy/cle/blob/3d17c5d6b648a4b14df66f1df505c9b3c37881a5/cle/backends/elf/elf.py#L548), so I was just matching it. It does feel a bit wonky since the module's name is the arch.

sraboy commented:
Alrighty, I've moved all the relocations under their backends. All the questions/reviews from above are resolved, I think.

I'm going to go ahead and start pulling out the ELF-specific stuff from Backend and Relocation.

sraboy commented:
Just wanted to provide a quick update: I've been running solely on this branch for the week and haven't run into any issues (though I've been using CLE on its own rather than angr, so I'm not as sure about integration). I'll be digging back in to a few binaries with angr this next week so I'll probably knock off a couple more of these todo's then.

rhelmot commented:
If you're doing heavy work on windows you might also want to use [my additions to pefile](https://github.com/erocarrera/pefile/pull/206), just as a heads up.

sraboy commented:
Okay, I rebased on the latest few changes and squashed all the old commits. I think I caught all the merge issues. I figured I'd wait on adding/updating the few other things I'd like to do until this refactor is merged in so I don't have to deal with the merge hell again.

Any more issues or concerns on this?

ltfish commented:
Travis CI should not fail now :-)

sraboy commented:
Ha, fixed just in time for me to come in and try to break things. I'll work on that.

rhelmot commented:
Almost there - just looks like you broke something related to MIPS, and one of the examples in the docs started failing.

sraboy commented:
There's also test_accuracy.test_arrays... having a hell of a time figuring that out.

sraboy commented:
Turns out some of those relocation modules got whacked when I rebased. Finally all set.

sraboy commented:
I rebased on the latest changes to PE and Travis broke with a timeout. Now it's breaking on the tracer stuff. I don't think it's related to this...

rhelmot commented:
Nah, its's just that sometimes travis flakes out and some http requests fail. I hit retry and it worked just fine.

Thank you so much!!!
